### PR TITLE
MM-20558: Migrate admin_console/plugin_management to Typescript

### PR DESCRIPTION
**Pull request by sowmiyamuthuraman**
Fixes https://github.com/mattermost/mattermost-server/issues/13514

JIRA: https://mattermost.atlassian.net/browse/MM-20558

### CODE REVIEW

**Comment by devinbinnie** [comment]
@sowmiyamuthuraman I think you'll want to look into the `Partial` helper type in TypeScript. It should help you solve a couple of the errors in here.

Here's a link to the documentation on `Partial`: https://www.typescriptlang.org/docs/handbook/advanced-types.html#mapped-types

**Comment by sowmiyamuthuraman** [comment]
@hanzei I have fixed the errors!

File: components/admin_console/plugin_management/plugin_management.tsx [hunk 1]
```diff
@@ -367,33 +383,63 @@ const PluginItem = ({
...
-PluginItem.propTypes = {
-    pluginStatus: PropTypes.object.isRequired,
-    removing: PropTypes.bool.isRequired,
-    handleEnable: PropTypes.func.isRequired,
-    handleDisable: PropTypes.func.isRequired,
-    handleRemove: PropTypes.func.isRequired,
-    showInstances: PropTypes.bool.isRequired,
-    hasSettings: PropTypes.bool.isRequired,
+type Props = BaseProps & {
+    config: {
```
Comment (nevyangelova) [inline components/admin_console/plugin_management/plugin_management.tsx:387]: I would extract this to a separate Interface especially if we pluginSettings can come from somewhere else and we would need to share that type.

File: components/admin_console/plugin_management/plugin_management.tsx [hunk 2]
```diff
@@ -165,7 +181,7 @@ const PluginItem = ({
         activateButton = (
             <a
                 data-plugin-id={pluginStatus.id}
-                disabled={deactivating}
+                data-disabled={deactivating}
```
Comment (devinbinnie) [inline components/admin_console/plugin_management/plugin_management.tsx:184]: Why were these changed?

File: components/admin_console/plugin_management/plugin_management.tsx (hunk 1 above)
Comment (devinbinnie) [reply components/admin_console/plugin_management/plugin_management.tsx:387]: We actually have a type for the `Config` that they can use in `mattermost-redux`

File: components/admin_console/plugin_management/plugin_management.tsx [hunk 3]
```diff
@@ -367,33 +383,63 @@ const PluginItem = ({
...
+            EnableRemoteMarketplace: boolean;
+            AutomaticPrepackagedPlugins: boolean;
+            MarketplaceUrl: string;
+            RequirePluginSignature: boolean;
+        };
+        ExperimentalSettings: {
+            RestrictSystemAdmin: boolean;
+        };
+    };
+    pluginStatuses: object;
```
Comment (devinbinnie) [inline components/admin_console/plugin_management/plugin_management.tsx:402]: Shouldn't this be of type `PluginStatus[]` that was defined above?

File: components/admin_console/plugin_management/plugin_management.tsx [hunk 4]
```diff
@@ -413,8 +459,7 @@ export default class PluginManagement extends AdminSettings {
             resolveRemoveModal: null,
         });
     }
-
-    getConfigFromState = (config) => {
+    getConfigFromState = (config: Props['config']) => {
```
Comment (devinbinnie) [inline components/admin_console/plugin_management/plugin_management.tsx:462]: Same here for the `Config` type

File: components/admin_console/plugin_management/plugin_management.tsx [hunk 5]
```diff
@@ -646,35 +694,40 @@ export default class PluginManagement extends AdminSettings {
 
     handleRemove = async () => {
         this.setState({lastMessage: null, serverError: null});
-        const {error} = await this.props.actions.removePlugin(this.state.removing);
-        this.setState({removing: null});
+        if (this.state.removing) {
```
Comment (devinbinnie) [inline components/admin_console/plugin_management/plugin_management.tsx:697]: Is this just a null check? This might be changing logic since it's also checking if it evaluates to true as well.

File: components/admin_console/plugin_management/plugin_management.tsx [hunk 6]
```diff
@@ -722,7 +777,9 @@ export default class PluginManagement extends AdminSettings {
         );
     }
 
-    renderRemovePluginModal = ({show, onConfirm, onCancel}) => {
+    renderRemovePluginModal = (
+        {show, onConfirm, onCancel}:
```
Comment (devinbinnie) [inline components/admin_console/plugin_management/plugin_management.tsx:781]: How come we need both these lines? The one below should take care of the typing.

File: components/admin_console/plugin_management/plugin_management.tsx (hunk 2 above)
Comment (sowmiyamuthuraman) [reply components/admin_console/plugin_management/plugin_management.tsx:184]: These were added because of the ts error `Type '{ children: Element; data-plugin-id: string; disabled: boolean; onClick: (e: any) => any; }' is not assignable to type 'DetailedHTMLProps<AnchorHTMLAttributes<HTMLAnchorElement>, HTMLAnchorElement>'.
  Property 'disabled' does not exist on type 'DetailedHTMLProps<AnchorHTMLAttributes<HTMLAnchorElement>, HTMLAnchorElement>'.ts(2322)`

File: components/admin_console/plugin_management/plugin_management.tsx (hunk 4 above)
Comment (sowmiyamuthuraman) [reply components/admin_console/plugin_management/plugin_management.tsx:462]: @devinbinnie  should i refer to this file `mattermost-redux/types/config.d.ts`? I found that the keys that are required for this file are not available in the *Config* type

Comment (devinbinnie) [reply components/admin_console/plugin_management/plugin_management.tsx:462]: Yes that's the correct type, which keys are missing? We can add those fields to the redux type.

File: components/admin_console/plugin_management/plugin_management.tsx (hunk 2 above)
Comment (devinbinnie) [reply components/admin_console/plugin_management/plugin_management.tsx:184]: It looks like `disabled` doesn't do anything on anchor tags, so we can actually just remove this line entirely. `data-disabled` would only be relevant if some code was consuming it.

File: components/admin_console/plugin_management/plugin_management.tsx (hunk 4 above)
Comment (sowmiyamuthuraman) [reply components/admin_console/plugin_management/plugin_management.tsx:459]: **PluginSettings**, **ExperimentalSettings** are the missing keys

`config: {
        PluginSettings: {
            Enable: boolean;
            EnableUploads: boolean;
            AllowInsecureDownloadUrl: boolean;
            EnableMarketplace: boolean;
            EnableRemoteMarketplace: boolean;
            AutomaticPrepackagedPlugins: boolean;
            MarketplaceUrl: string;
            RequirePluginSignature: boolean;
        };
        ExperimentalSettings: {
            RestrictSystemAdmin: boolean;
        };
    };
`

File: components/admin_console/plugin_management/plugin_management.tsx (hunk 1 above)
Comment (nevyangelova) [reply components/admin_console/plugin_management/plugin_management.tsx:384]: @sowmiyamuthuraman thanks for handling. Have you had a chance to consider my comment here?

File: components/admin_console/plugin_management/plugin_management.tsx (hunk 4 above)
Comment (sowmiyamuthuraman) [reply components/admin_console/plugin_management/plugin_management.tsx:459]: @sowmiyamuthuraman thanks for handling. Have you had a chance to consider my comment here?
@nevyangelova some of the keys are missing in the config type. Waiting for reply from @devinbinnie

File: components/admin_console/plugin_management/plugin_management.tsx (hunk 1 above)
Comment (nevyangelova) [reply components/admin_console/plugin_management/plugin_management.tsx:384]: > I would extract this to a separate Interface especially if we pluginSettings can come from somewhere else and we would need to share that type.

@sowmiyamuthuraman you can still extract this to a separate interface if keys are missing, so we can reuse and share the type :)

Comment (sowmiyamuthuraman) [reply components/admin_console/plugin_management/plugin_management.tsx:395]: @nevyangelova I have made the changes. But the build is failing due to`sh: 1: eslint: not found`. can you pls help?

**Comment by nevyangelova** [comment]
@sowmiyamuthuraman looks like you are missing package.json, this happens sometimes with forked repos. I will try merging master (in some cases this fixes it) and if not, you might have to run npm install locally and push again. Thanks for the contribution 💯

File: components/admin_console/plugin_management/plugin_management.tsx (hunk 4 above)
Comment (devinbinnie) [reply components/admin_console/plugin_management/plugin_management.tsx:460]: Hey sorry for the late reply @sowmiyamuthuraman, I've been preoccupied.

Just spent some time digging through the `admin_settings` code, and for some reason they're using a different type definition for the `Config` object than `mattermost-redux`

This feels like something we'll need to cleanup separately, so we can make use of the types you've already defined for this conversion. Thanks for all your work on this :)

File: components/admin_console/plugin_management/plugin_management.tsx [hunk 7]
```diff
@@ -452,13 +495,13 @@ export default class PluginManagement extends AdminSettings {
 
     handleUpload = () => {
         this.setState({lastMessage: null, serverError: null});
-        const element = this.refs.fileInput;
-        if (element.files.length > 0) {
+        const element = this.refs.fileInput as HTMLInputElement;
+        if (element.files && element.files.length > 0) {
```
Comment (devinbinnie) [inline components/admin_console/plugin_management/plugin_management.tsx:499]: You can use optional chaining here:
```suggestion
        if (element.files?.length > 0) {
```

Comment (sowmiyamuthuraman) [reply components/admin_console/plugin_management/plugin_management.tsx:499]: @devinbinnie optional chaining is throwing type error.`TS2532: Object is possibly 'undefined'.`

Comment (devinbinnie) [reply components/admin_console/plugin_management/plugin_management.tsx:499]: Odd, we fixed that...leave it as it I suppose. Thanks again :)

**Comment by jasonblais** [comment]
@devinbinnie @nevyangelova does this PR require QA review? From reading the ticket, it sounds like no. If so, this PR would be ready to merge after conflicts are resolved.

@sowmiyamuthuraman Really apologize for such a long delay in responding. Would you be open to help resolve the merge conflicts? Many many thanks in advance..

**Comment by srkgupta** [comment]
Requesting review from @prapti since this involves changes related to Plugins. @prapti can you please do some sanity testing on the System console Plugins page and then approve the PR if its working fine. Thanks.

**Comment by hanzei** [comment]
@prapti Gentle reminder to help with the review of this PR :)

**Comment by hanzei** [comment]
Thanks for the great contribution @sowmiyamuthuraman :+1:
